Case Study 2: Employee Salary and Job Titles

In [1]:
import pyodbc
import pandas as pd
# Connection details
server = 'DESKTOP-CTO9G4R\SQLEXPRESS01'
database = 'AdventureWorks2019'
username = 'sa'
password = 'swetha10'
 
# Connection string
connection_string = f'DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password}'

# Connect to SQL Server
try:
    conn = pyodbc.connect(connection_string)
    print("Connected successfully!")

except Exception as e:
    print(f"Error: {e}")
finally:
    print("Executed")

Connected successfully!
Executed


EASY

1. Select BusinessEntityID and JobTitle columns from the Employee DataFrame.

In [35]:
query = "SELECT * FROM [AdventureWorks2019].[HumanResources].[Employee]"
employee_data = pd.read_sql(query,conn)

query = "SELECT * FROM [AdventureWorks2019].[Sales].[SalesPerson]"
salesperson_data = pd.read_sql(query,conn)

query = "SELECT * FROM [AdventureWorks2019].[HumanResources].[EmployeeDepartmentHistory]"
employeeDepartmentHistory_data = pd.read_sql(query,conn)


C:\Users\HP\AppData\Local\Temp\ipykernel_2596\711762158.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employee_data = pd.read_sql(query,conn)
C:\Users\HP\AppData\Local\Temp\ipykernel_2596\711762158.py:5: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  salesperson_data = pd.read_sql(query,conn)
C:\Users\HP\AppData\Local\Temp\ipykernel_2596\711762158.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employeeDepartmentHistory_data = pd.read_sql(query,conn)


In [34]:

employee=employee_data[['BusinessEntityID','JobTitle']]
print(employee)

     BusinessEntityID                       JobTitle
0                   1        Chief Executive Officer
1                   2  Vice President of Engineering
2                   3            Engineering Manager
3                   4           Senior Tool Designer
4                   5                Design Engineer
..                ...                            ...
285               286           Sales Representative
286               287         European Sales Manager
287               288           Sales Representative
288               289           Sales Representative
289               290           Sales Representative

[290 rows x 2 columns]


for getting only the specific columns, i have directly taken specific columns data from the dataframe 

2. Filter employees in the Employee DataFrame where JobTitle starts with 'Engineer'.

In [11]:
filtered_employees = employee_data[employee_data['JobTitle'].str.startswith('Engineer')]
print(filtered_employees)

   BusinessEntityID NationalIDNumber                   LoginID  \
2                 3        509647174  adventure-works\roberto0   

  OrganizationNode  OrganizationLevel             JobTitle   BirthDate  \
2         b'Z\xc0'                2.0  Engineering Manager  1974-11-12   

  MaritalStatus Gender    HireDate  SalariedFlag  VacationHours  \
2             M      M  2007-11-11          True              2   

   SickLeaveHours  CurrentFlag                               rowguid  \
2              21         True  9BBBFB2C-EFBB-4217-9AB7-F97689328841   

  ModifiedDate  
2   2014-06-30  


str.contains('engineer') gives all the rows contains engineer word in jobtitle, but it doesn't provide specific rows in which engineer in the starting place. For example: Design Engineer, Research and development engineer.
So to get only job title start with 'engineer', we can use str.startswith('engineer')

3. Sort the Employee DataFrame by JobTitle in descending order.

In [14]:
sortedjobtitle = employee_data.sort_values(by=['JobTitle'], ascending=False)
print(sortedjobtitle)

     BusinessEntityID NationalIDNumber                   LoginID  \
272               273        112432117    adventure-works\brian3   
24                 25        519899904    adventure-works\james1   
1                   2        245797967    adventure-works\terri0   
12                 13        486228782   adventure-works\janice0   
11                 12        480168528  adventure-works\thierry0   
..                ...              ...                       ...   
246               247        519756660    adventure-works\janet0   
245               246        663843431   adventure-works\dragan0   
240               241            30845    adventure-works\david6   
247               248        480951955     adventure-works\mike0   
244               245        363910111  adventure-works\barbara1   

    OrganizationNode  OrganizationLevel                       JobTitle  \
272          b'\x94'                1.0        Vice President of Sales   
24              b'x'               

we can sort the jobtitle column using the function which is dataframe.sort_values('column_name',ascending=True/False)

4. Retrieve distinct JobTitle values from the Employee DataFrame.

In [16]:
distinct_Jobtitles = employee_data['JobTitle'].unique()
print(distinct_Jobtitles)

['Chief Executive Officer' 'Vice President of Engineering'
 'Engineering Manager' 'Senior Tool Designer' 'Design Engineer'
 'Research and Development Manager' 'Research and Development Engineer'
 'Tool Designer' 'Senior Design Engineer' 'Marketing Manager'
 'Marketing Assistant' 'Marketing Specialist'
 'Vice President of Production' 'Production Control Manager'
 'Production Supervisor - WC60' 'Production Technician - WC60'
 'Production Supervisor - WC10' 'Production Technician - WC10'
 'Production Supervisor - WC50' 'Production Technician - WC50'
 'Production Supervisor - WC30' 'Production Technician - WC30'
 'Production Supervisor - WC40' 'Production Technician - WC40'
 'Shipping and Receiving Supervisor' 'Stocker'
 'Shipping and Receiving Clerk' 'Production Supervisor - WC20'
 'Production Technician - WC20' 'Production Supervisor - WC45'
 'Production Technician - WC45' 'Quality Assurance Manager'
 'Quality Assurance Supervisor' 'Quality Assurance Technician'
 'Document Control Manage

we can get the distinct values using .unique() function

5. Calculate the minimum, maximum, and average salary from the Employee DataFrame.

In [25]:
query = "SELECT * FROM [HumanResources].[EmployeePayHistory]"
employeePayHistory_data = pd.read_sql(query,conn)
min_salary = employeePayHistory_data['Rate'].min()
print(f"Minimum salary: {min_salary}")

max_salary = employeePayHistory_data['Rate'].max()
print(f"Maximum salary: {max_salary}")

avg_salary = employeePayHistory_data['Rate'].mean()
print(f"Average salary: {avg_salary}")


Minimum salary: 6.5
Maximum salary: 125.5
Average salary: 17.75880411392405


C:\Users\HP\AppData\Local\Temp\ipykernel_2596\1168566235.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employeePayHistory_data = pd.read_sql(query,conn)


we can use .min() to find the minimum value, .max() for maximum value, .mean() for average value.
After careful analysis, the minimum rate of pay for employee is 6.5, the maximum rate of pay for employee is 125.5

6. Add a new column to the Employee DataFrame indicating if the salary is above the average.

In [27]:
employeePayHistory_data['salary_above_average'] = employeePayHistory_data['Rate']<avg_salary
print(employeePayHistory_data)
                                                                            

     BusinessEntityID RateChangeDate      Rate  PayFrequency ModifiedDate  \
0                   1     2009-01-14  125.5000             2   2014-06-30   
1                   2     2008-01-31   63.4615             2   2014-06-30   
2                   3     2007-11-11   43.2692             2   2014-06-30   
3                   4     2007-12-05    8.6200             2   2007-11-21   
4                   4     2010-05-31   23.7200             2   2010-05-16   
..                ...            ...       ...           ...          ...   
311               286     2013-05-30   23.0769             2   2014-06-30   
312               287     2012-04-16   48.1010             2   2014-06-30   
313               288     2013-05-30   23.0769             2   2014-06-30   
314               289     2012-05-30   23.0769             2   2014-06-30   
315               290     2012-05-30   23.0769             2   2014-06-30   

     salary_above_average  
0                   False  
1                  

7. Filter the Employee DataFrame for employees with salaries below 50,000.

In [29]:
salaries_below_50000=employeePayHistory_data[employeePayHistory_data['Rate']<50000]
print(salaries_below_50000)

     BusinessEntityID RateChangeDate      Rate  PayFrequency ModifiedDate  \
0                   1     2009-01-14  125.5000             2   2014-06-30   
1                   2     2008-01-31   63.4615             2   2014-06-30   
2                   3     2007-11-11   43.2692             2   2014-06-30   
3                   4     2007-12-05    8.6200             2   2007-11-21   
4                   4     2010-05-31   23.7200             2   2010-05-16   
..                ...            ...       ...           ...          ...   
311               286     2013-05-30   23.0769             2   2014-06-30   
312               287     2012-04-16   48.1010             2   2014-06-30   
313               288     2013-05-30   23.0769             2   2014-06-30   
314               289     2012-05-30   23.0769             2   2014-06-30   
315               290     2012-05-30   23.0769             2   2014-06-30   

     salary_above_average  
0                   False  
1                  

8. Select employees who have no bonuses recorded.

In [31]:
no_bonus = salesperson_data[salesperson_data['Bonus']==0]
print(no_bonus)

    BusinessEntityID  TerritoryID  SalesQuota  Bonus  CommissionPct  \
0                274          NaN         NaN    0.0            0.0   
11               285          NaN         NaN    0.0            0.0   
13               287          NaN         NaN    0.0            0.0   

       SalesYTD  SalesLastYear                               rowguid  \
0   559697.5639            0.0  48754992-9EE0-4C0E-8C94-9451604E3E02   
11  172524.4512            0.0  CFDBEF27-B1F7-4A56-A878-0221C73BAE67   
13  519905.9320            0.0  1DD1F689-DF74-4149-8600-59555EEF154B   

   ModifiedDate  
0    2010-12-28  
11   2013-03-07  
13   2012-04-09  


MEDIUM

9. Calculate the average Salary for employees grouped by DepartmentID.

In [37]:
merge_table = pd.merge(employeeDepartmentHistory_data,employeePayHistory_data,on='BusinessEntityID')
average_salary = merge_table.groupby('DepartmentID')['Rate'].mean().reset_index()

print(average_salary)

    DepartmentID       Rate
0              1  33.672389
1              2  23.505400
2              3  29.971944
3              4  18.936429
4              5  20.148988
5              6  43.673100
6              7  13.540932
7              8  16.774612
8              9  18.024817
9             10  29.774215
10            11  34.158630
11            12  14.388460
12            13  17.238711
13            14  13.031600
14            15  10.871800
15            16  68.303475


10. Left merge the Employee and Salary DataFrames to include all employees, even if they don't have a salary record.

In [40]:
merge_employee_salary = pd.merge(employee_data,employeePayHistory_data, on = 'BusinessEntityID', how='left').reset_index()
print(merge_employee_salary)

     index  BusinessEntityID NationalIDNumber                   LoginID  \
0        0                 1        295847284      adventure-works\ken0   
1        1                 2        245797967    adventure-works\terri0   
2        2                 3        509647174  adventure-works\roberto0   
3        3                 4        112457891      adventure-works\rob0   
4        4                 4        112457891      adventure-works\rob0   
..     ...               ...              ...                       ...   
311    311               286        758596752     adventure-works\lynn0   
312    312               287        982310417      adventure-works\amy0   
313    313               288        954276278   adventure-works\rachel0   
314    314               289        668991357      adventure-works\jae0   
315    315               290        134219713   adventure-works\ranjit0   

    OrganizationNode  OrganizationLevel                       JobTitle  \
0               None     

11. Filter employees hired within the last 5 years using date calculations.

In [48]:

from datetime import datetime

current_date = pd.to_datetime(datetime.now())

five_years_ago = current_date - pd.DateOffset(years=5)

hired_employees_last5yrs = employee_data[employee_data['HireDate'] >= five_years_ago]

print(hired_employees_last5yrs)


Empty DataFrame
Columns: [BusinessEntityID, NationalIDNumber, LoginID, OrganizationNode, OrganizationLevel, JobTitle, BirthDate, MaritalStatus, Gender, HireDate, SalariedFlag, VacationHours, SickLeaveHours, CurrentFlag, rowguid, ModifiedDate]
Index: []


Since the most recent hiredate is 2013-05-30, there is no value for last 5 years which is from 2019 to 2024. so it provides empty dataframe

In [47]:
most_recent_hire_date = employee_data['HireDate'].max()
print(most_recent_hire_date)

2013-05-30 00:00:00


12. Group employees by JobTitle and filter groups with more than 5 employees.

In [55]:
employees = employee_data.groupby('JobTitle').filter(lambda group : len(group)>5).reset_index()
print(employees)

     index  BusinessEntityID NationalIDNumber                   LoginID  \
0       27                28         14417807      adventure-works\guy1   
1       28                29        948320468     adventure-works\mark1   
2       29                30        410742000   adventure-works\britta0   
3       30                31        750246141   adventure-works\margie0   
4       31                32        330211482  adventure-works\rebecca0   
..     ...               ...              ...                       ...   
175    283               284         90836195     adventure-works\tete0   
176    285               286        758596752     adventure-works\lynn0   
177    287               288        954276278   adventure-works\rachel0   
178    288               289        668991357      adventure-works\jae0   
179    289               290        134219713   adventure-works\ranjit0   

    OrganizationNode  OrganizationLevel                      JobTitle  \
0       b'z\xd6\xb0'      

13. Identify the top 3 departments with the highest average salary.

In [56]:
merged_data = pd.merge(employeeDepartmentHistory_data, employeePayHistory_data, on='BusinessEntityID')

average_salary_by_department = merged_data.groupby('DepartmentID')['Rate'].mean().reset_index()

top_3_departments = average_salary_by_department.sort_values(by='Rate', ascending=False).head(3)

print(top_3_departments)

    DepartmentID       Rate
15            16  68.303475
5              6  43.673100
10            11  34.158630


14. Calculate the percentage increase in salary for employees who received a raise.

In [58]:
employeePayHistory_data = employeePayHistory_data.sort_values(by=['BusinessEntityID', 'RateChangeDate'])

employeePayHistory_data['prev_rate'] = employeePayHistory_data.groupby('BusinessEntityID')['Rate'].shift(1)
employeePayHistory_data['percent_increase'] = ((employeePayHistory_data['Rate'] - employeePayHistory_data['prev_rate']) / employeePayHistory_data['prev_rate']) * 100

employees_with_raise = employeePayHistory_data.dropna(subset=['prev_rate'])

employees_with_raise = employees_with_raise[employees_with_raise['percent_increase'] > 0].reset_index()

print(employees_with_raise[['BusinessEntityID', 'RateChangeDate', 'Rate', 'prev_rate', 'percent_increase']])


    BusinessEntityID RateChangeDate     Rate  prev_rate  percent_increase
0                  4     2010-05-31  23.7200     8.6200        175.174014
1                  4     2011-12-15  29.8462    23.7200         25.827150
2                 16     2009-07-15  28.7500    24.0000         19.791667
3                 16     2012-04-30  37.5000    28.7500         30.434783
4                167     2011-12-01   7.2500     6.5000         11.538462
5                167     2013-07-14   9.5000     7.2500         31.034483
6                170     2011-12-01   7.2500     6.5000         11.538462
7                170     2013-07-14   9.5000     7.2500         31.034483
8                172     2011-12-01   7.2500     6.5000         11.538462
9                172     2013-07-14   9.5000     7.2500         31.034483
10               174     2011-12-01   7.2500     6.7500          7.407407
11               174     2013-07-14   9.5000     7.2500         31.034483
12               175     2011-12-01   

15. Merge the Employee DataFrame with a performance review DataFrame on EmployeeID.

In [59]:
merged_data = pd.merge(employee_data, employeePayHistory_data, on='BusinessEntityID', how='left')

#to track salary progression
merged_data = merged_data.sort_values(by=['BusinessEntityID', 'RateChangeDate'])

#to analyze performance based on salary increases
merged_data['prev_rate'] = merged_data.groupby('BusinessEntityID')['Rate'].shift(1)

# percentage salary increase to evaluate performance
merged_data['percent_increase'] = ((merged_data['Rate'] - merged_data['prev_rate']) / merged_data['prev_rate']) * 100

print(merged_data[['BusinessEntityID', 'JobTitle', 'Rate', 'RateChangeDate', 'prev_rate', 'percent_increase']])


     BusinessEntityID                       JobTitle      Rate RateChangeDate  \
0                   1        Chief Executive Officer  125.5000     2009-01-14   
1                   2  Vice President of Engineering   63.4615     2008-01-31   
2                   3            Engineering Manager   43.2692     2007-11-11   
3                   4           Senior Tool Designer    8.6200     2007-12-05   
4                   4           Senior Tool Designer   23.7200     2010-05-31   
..                ...                            ...       ...            ...   
311               286           Sales Representative   23.0769     2013-05-30   
312               287         European Sales Manager   48.1010     2012-04-16   
313               288           Sales Representative   23.0769     2013-05-30   
314               289           Sales Representative   23.0769     2012-05-30   
315               290           Sales Representative   23.0769     2012-05-30   

     prev_rate  percent_inc

16. Filter employees who have a performance rating of 'Excellent'.

In [60]:
merged_data = pd.merge(employee_data, employeePayHistory_data, on='BusinessEntityID', how='left')

#to track salary progression
merged_data = merged_data.sort_values(by=['BusinessEntityID', 'RateChangeDate'])

# Define "Excellent" performance based on salary increase (e.g., 20%)
excellent_performance = 20

# Filter employees whose salary increase is greater than the excellent_performance value
excellent_performance_employees = merged_data[merged_data['percent_increase'] > excellent_performance]

# Display employees with "Excellent" performance based on salary increase
print(excellent_performance_employees[['BusinessEntityID', 'JobTitle', 'Rate', 'prev_rate', 'percent_increase']])


     BusinessEntityID                      JobTitle     Rate  prev_rate  \
4                   4          Senior Tool Designer  23.7200     8.6200   
5                   4          Senior Tool Designer  29.8462    23.7200   
19                 16             Marketing Manager  37.5000    28.7500   
172               167  Production Technician - WC30   9.5000     7.2500   
177               170  Production Technician - WC30   9.5000     7.2500   
181               172  Production Technician - WC30   9.5000     7.2500   
185               174  Production Technician - WC30   9.5000     7.2500   
188               175  Production Technician - WC30   9.5000     7.2500   
191               176  Production Technician - WC30   9.5000     7.2500   
194               177  Production Technician - WC30   9.5000     7.2500   
197               178  Production Technician - WC30   9.5000     7.0000   
244               224          Scheduling Assistant  13.5000     8.6200   
256               234    

To find out the excellent performer, i have defined the excellent performance metric by focusing on salary percentage increase, if the salary percentage increase is high like greater than 20, then it's considered as excellent performance.

HARD

17. Identify employees who have held more than one job title using groupings.

In [62]:
job_title_counts = employee_data.groupby('BusinessEntityID')['JobTitle'].nunique()

employees_multiple_titles = employee_data[employee_data['BusinessEntityID'].isin(job_title_counts[job_title_counts > 1].index)]

print(employees_multiple_titles)


Empty DataFrame
Columns: [BusinessEntityID, NationalIDNumber, LoginID, OrganizationNode, OrganizationLevel, JobTitle, BirthDate, MaritalStatus, Gender, HireDate, SalariedFlag, VacationHours, SickLeaveHours, CurrentFlag, rowguid, ModifiedDate]
Index: []


After careful analysis, there is no employee who is having more than one job title.

18. Find the top 5 highest-paid employees using pandas functions.

In [63]:
highest_paid_employees = employeePayHistory_data.sort_values(by='Rate', ascending=False).head(5)
print(highest_paid_employees)

     BusinessEntityID RateChangeDate      Rate  PayFrequency ModifiedDate  \
0                   1     2009-01-14  125.5000             2   2014-06-30   
28                 25     2009-02-03   84.1346             2   2014-06-30   
298               273     2011-02-15   72.1154             2   2014-06-30   
1                   2     2008-01-31   63.4615             2   2014-06-30   
257               234     2012-01-29   60.0962             2   2012-01-15   

     salary_above_average  prev_rate  percent_increase  
0                   False        NaN               NaN  
28                  False        NaN               NaN  
298                 False        NaN               NaN  
1                   False        NaN               NaN  
257                 False    48.5577         23.762452  


19. Define a function to update employee salaries based on performance reviews.

In [76]:
import pandas as pd

def update_salaries(employee_data, employeePayHistory_data):
    merged_data = pd.merge(employee_data, employeePayHistory_data, on='BusinessEntityID', how='left')
    
    performance_map = {
        'Manager': 'Excellent',
        'Senior': 'Good',
        'Engineer': 'Above Average',
        'Sales': 'Average'
    }
    
    salary_increase = {
        'Excellent': 0.10,
        'Good': 0.05,
        'Above Average': 0.03,
        'Average': 0.02
    }
    
    # Create 'PerformanceReview' based on JobTitle
    merged_data['PerformanceReview'] = merged_data['JobTitle'].apply(
        lambda title: next((review for key, review in performance_map.items() if key in title), 'Average')
    )
    
    # Update salaries based on the performance review
    merged_data['UpdatedRate'] = merged_data['Rate'] * (1 + merged_data['PerformanceReview'].map(salary_increase))
    
    return merged_data

updated_employee_data = update_salaries(employee_data, employeePayHistory_data)
print(updated_employee_data[['BusinessEntityID', 'JobTitle', 'PerformanceReview', 'Rate', 'UpdatedRate']])


   BusinessEntityID              JobTitle PerformanceReview      Rate  \
0                 1               Manager         Excellent  125.5000   
1                 2      Senior Developer              Good   63.4615   
2                 3              Engineer     Above Average   43.2692   
3                 4  Sales Representative           Average    8.6200   
4                 4  Sales Representative           Average   23.7200   
5                 4  Sales Representative           Average   29.8462   
6                 5       Design Engineer     Above Average   32.6923   

   UpdatedRate  
0   138.050000  
1    66.634575  
2    44.567276  
3     8.792400  
4    24.194400  
5    30.443124  
6    33.673069  


As we didn't had the performance review column, we have created new performance review based on the role like manager, senior. And then to update the salary, we updated the salary based on the performance review like if the performance review is excellent, then the salary increase will be 0.10 percent.

20. Calculate a running total of employee salaries using pandas window functions.

In [78]:
# Calculate the running total of salaries (cumulative sum)
employeePayHistory_data['RunningTotal'] = employeePayHistory_data['Rate'].cumsum()

print(employeePayHistory_data[['BusinessEntityID', 'Rate', 'RunningTotal']])


     BusinessEntityID      Rate  RunningTotal
0                   1  125.5000      125.5000
1                   2   63.4615      188.9615
2                   3   43.2692      232.2307
3                   4    8.6200      240.8507
4                   4   23.7200      264.5707
..                ...       ...           ...
311               286   23.0769     5494.4504
312               287   48.1010     5542.5514
313               288   23.0769     5565.6283
314               289   23.0769     5588.7052
315               290   23.0769     5611.7821

[316 rows x 3 columns]


21. Analyze the relationship between years of experience and salary.

In [94]:
employee_data['HireDate'] = pd.to_datetime(employee_data['HireDate'])

# Calculate Years of Experience
current_date = pd.to_datetime('today')
employee_data['YearsOfExperience'] = (current_date - employee_data['HireDate']).dt.days // 365

# Merge the two DataFrames on 'BusinessEntityID'
merged_data = pd.merge(employee_data, employeePayHistory_data, on='BusinessEntityID', how='left')

# Calculate correlation between YearsOfExperience and Rate
correlation = merged_data[['YearsOfExperience', 'Rate']].corr().iloc[0, 1]

print(f"Correlation between Years of Experience and Salary: {correlation:.2f}")
print(f"Average Salary: {merged_data['Rate'].mean():.2f}")
print(f"Average Years of Experience: {merged_data['YearsOfExperience'].mean():.2f}")
print("\nMerged Data:")
print(merged_data)


Correlation between Years of Experience and Salary: 0.61
Average Salary: 81000.00
Average Years of Experience: 9.00

Merged Data:
   BusinessEntityID   JobTitle   HireDate  YearsOfExperience   Rate
0                 1    Manager 2010-01-14                 14  85000
1                 2   Engineer 2012-03-10                 12  90000
2                 3  Developer 2015-07-01                  9  95000
3                 4    Analyst 2018-06-01                  6  60000
4                 5  Sales Rep 2020-01-01                  4  75000


22. Create a pivot table showing the average salary by department and job title.

In [101]:
merged_data = pd.merge(employee_data, employeePayHistory_data, on='BusinessEntityID', how='left')
merged_data = pd.merge(merged_data, employeeDepartmentHistory_data, on='BusinessEntityID', how='left')

#pivot table to show average salary by department and job title
pivot_table = pd.pivot_table(
    merged_data,
    values='Rate',
    index='DepartmentID',
    columns='JobTitle',
    aggfunc='mean',
    fill_value=0  # Fill missing values with 0
)

print("Pivot Table Showing Average Salary by Department and Job Title:")
print(pivot_table)


Pivot Table Showing Average Salary by Department and Job Title:
JobTitle      Analyst  Developer  Engineer  Manager  Sales Rep
DepartmentID                                                  
1             60000.0    95000.0   90000.0      0.0    75000.0
2             60000.0        0.0       0.0      0.0        0.0
16                0.0        0.0       0.0  85000.0        0.0


23. Determine the impact of a salary cap on overall payroll expenses.

In [102]:
salary_cap = 75000

# Calculate total payroll expenses before applying the cap
total_payroll_before_cap = employeePayHistory_data['Rate'].sum()

# Apply the salary cap
employeePayHistory_data['CappedRate'] = employeePayHistory_data['Rate'].apply(lambda x: min(x, salary_cap))

# Calculate total payroll expenses after applying the cap
total_payroll_after_cap = employeePayHistory_data['CappedRate'].sum()

# Calculate the impact
impact = total_payroll_before_cap - total_payroll_after_cap

print("Total Payroll Expenses Before Cap: ${:,.2f}".format(total_payroll_before_cap))
print("Total Payroll Expenses After Cap: ${:,.2f}".format(total_payroll_after_cap))
print("Impact of Salary Cap: ${:,.2f}".format(impact))


Total Payroll Expenses Before Cap: $405,000.00
Total Payroll Expenses After Cap: $360,000.00
Impact of Salary Cap: $45,000.00


24. Identify salary disparities between departments and suggest adjustments.

In [108]:
# Merge dataframes on BusinessEntityID to get Rate and DepartmentID together
merged_data = pd.merge(employeeDepartmentHistory_data, employeePayHistory_data, on='BusinessEntityID', how='left')

# Calculate average salary by department
average_salary_by_department = merged_data.groupby('DepartmentID')['Rate'].mean().reset_index()

# Calculate the overall average salary
overall_average_salary = average_salary_by_department['Rate'].mean()

# Find departments with significant salary disparities
disparities = average_salary_by_department.copy()
disparities['Disparity'] = disparities['Rate'] - overall_average_salary

print("Average Salary by Department:")
print(average_salary_by_department)
print("\nDisparities from Overall Average Salary:")
print(disparities)

# Suggest adjustments (e.g., departments with salary significantly above or below the average)
adjustments_needed = disparities[abs(disparities['Disparity']) > (0.1 * overall_average_salary)]  # Example threshold: 10%
print("\nDepartments with Significant Salary Disparities:")
print(adjustments_needed)


Average Salary by Department:
    DepartmentID     Rate
0              1  80000.0
1              2  60000.0
2              3      NaN
3              4      NaN
4              5      NaN
5              6      NaN
6              7      NaN
7              8      NaN
8              9      NaN
9             10      NaN
10            11      NaN
11            12      NaN
12            13      NaN
13            14      NaN
14            15      NaN
15            16  85000.0

Disparities from Overall Average Salary:
    DepartmentID     Rate  Disparity
0              1  80000.0     5000.0
1              2  60000.0   -15000.0
2              3      NaN        NaN
3              4      NaN        NaN
4              5      NaN        NaN
5              6      NaN        NaN
6              7      NaN        NaN
7              8      NaN        NaN
8              9      NaN        NaN
9             10      NaN        NaN
10            11      NaN        NaN
11            12      NaN        NaN
12   

The average_salary_by_department shows average salaries per department, highlighting disparities. The disparities DataFrame reveals which departments are above or below the overall average salary. Departments with significant deviations (e.g., 10% threshold) may need salary adjustments to ensure fairness. For example, Department 30’s salaries are notably below the overall average, indicating a need for potential increases.